src/Cardano/Wallet/CoinSelection/Fee.hs

```diff
@@ -0,0 +1,351 @@
+  where
+    go remaining as | remaining == Coin 0 =
+        as
+    go remaining [] = [remaining]
+        -- we only create new change if for whatever reason there is none already
+        -- or if is some overflow happens when we try to add.
+    go remaining [a] =
+        let newChange =  Coin $ (getCoin remaining) + (getCoin a)
+        in
+            (if isValidCoin newChange then [newChange] else [a, remaining])
+    go rest@(Coin remaining) ls@(a : as) =
+      let piece = fromIntegral remaining `div` (length ls)
+          newRemaining = Coin $ fromIntegral $ fromIntegral remaining - piece
+          newChange = Coin $ fromIntegral $ piece + ((fromIntegral . getCoin) a)
+      in
+          (if isValidCoin newChange then
+               newChange : go newRemaining as
+           else
+               a : go rest as)
+
+-- Computing actual fee is a bit tricky in the generic realm because we don't
+-- know what type representation is used by the underlying implementation. So,
+-- we can't just sum up all the input and substract the sub of all outputs
+-- (incl. change) because we'll risk an overflow with each sum. Instead, we
+-- reduce the input value iteratively, coin by coin using a safe distance
+-- between coins that are known to be within bounds.
+-- The algorithm converge because we know that by construction, there are less
+-- outputs than inputs. In essence, this computes:
+--
+--     fees = ∑ inputs - (∑ outputs + ∑ changes)
+computeFee
+    :: CoinSelection
+    -> Fee
+computeFee (CoinSelection inps outs chgs) =
+    Fee
+    $ collapse
+    (map (coin . snd) inps)
+    (filter (> Coin 0) $ (map coin outs) <> chgs)
+  where
+    collapse :: [Coin] -> [Coin] -> Coin
+    -- Some remaining inputs together. At this point, we've removed
+    -- all outputs and changes, so what's left are simply the actual fees.
+    -- It's unrealistic to imagine them being bigger than the max coin value.
+    collapse plus [] =
+        invariant
+        "fees are bigger than max coin value"
+        (Coin . sum $ map getCoin plus)
+        isValidCoin
+
+    -- In order to safely compute fees at this level, we need make sure we don't
+    -- overflow. Therefore, we remove outputs to inputs until there's no outputs
+    -- left to remove.
+    collapse (p:ps) (m:ms)
+        | p > m =
+              let p' = Coin $ distance (getCoin p) (getCoin m)
+              in collapse (p':ps) ms
+        | p < m =
+              let m' = Coin $ distance (getCoin p) (getCoin m)
+              in collapse ps (m':ms)
+        | otherwise = collapse ps ms
+
+    -- This branch can only happens if we've depleted all our inputs and there
+    -- are still some outputs left to remove from them. If means the total value
+    -- of outputs (incl. change) was bigger than the total input value which is
+    -- by definition, impossible; unless we messed up real hard.
+    collapse []  _ =
+        invariant "outputs are bigger than inputs" (Coin 0) (const True)
```

Comment:
I believe it should be `const False` here, and let's replace `Coin 0` with `undefined` here ... this can never be reached.